### ISSUE TITLE
Support unix-like URI /ipfs/ in address bar

### ISSUE THREAD

**Issue by lidel** [lidel/ipfs-firefox-addon#13]
For lazy copy&paste ;-)
It should work the same way as currently exiting `ipfs:`

**Comment by lidel** [comment]
Actually, Firefox redirects `/ipfs/<path>` to `file:///ipfs/<path>` by default.
The file path will contain IPFS resources (provided via fuse after  `ipfs mount`).
This feels much better than random hack of protocol handling in addressbar. Nevermind then :-)

**Comment by lidel** [comment]
On the second thought this only works on unix-like systems with arbitrary mount points. 

In the light of https://github.com/lidel/ipfs-firefox-addon/issues/16#issuecomment-92343901  add-on should intercept requests to  `file:///ipfs/*` and `file:///ipns/*`, check if file actually exists, and if it does not (eg. Windows, or fuse not being mounted), redirect to currently active HTTP gateway.

**Comment by dylanPowers** [comment]
@lidel I'm not sure how Firefox works but Chrome got very cranky for me when web apps load from a file directory and try to load another file. You may want to check one of the markdown readers to be sure that isn't the case. 
Check if this loads for instance /ipfs/QmSrCRJmzE4zE1nAfWPbzVfanKQNBhp7ZWmMnEdbiLvYNh/mdown#/ipfs/QmfQ75DjAxYzxMP2hdm6o4wFwZS5t7uorEZ2pX9AKXEg2u

**Comment by lidel** [comment]
Yes, I noticed this under Firefox too:
> ![2015-04-13-200414_734x432_scrot](https://cloud.githubusercontent.com/assets/157609/7121974/70f96e6e-e218-11e4-8cc2-60665412781a.png)

There are multiple issues with `file://` and my current plan is to always redirect
`file:///ipfs/..` to HTTP gateway (at least for now), as it will solve bulk of issues and give us single 'transport' to focus on in browser add-on.

Running JS from HTTP Gateway may leave us with some [CORS](https://en.wikipedia.org/wiki/CORS) limitations (eg. [under Firefox when using custom protocol handler](https://github.com/lidel/ipfs-firefox-addon/issues/3)), but this will eventually be solved by adding HTTP headers described in  https://github.com/ipfs/go-ipfs/issues/934#issuecomment-87410776.

**Comment by lidel** [comment]
Quick brain dump for my future self: standard addon SDK in Firefox does not provide good hook/event that could be used to replace `file:///` requests that end with 404 (file not found, when fuse is not mounted, which is the case on MS Windows when running http gateway on a different machine).  I don't want to listen for `*` events due to potential performance impact.

Replacing protocol handler for `file://` seems to be very invasive and from what I've needs to be done before some XUL elements are initialized during browser startup.  Not sure if it is possible to do it via Jetpack-based addon.

**tl;dr**:  support for `/ipfs/<hash>` in Firefox (without fuse-based mountpoint in operating system) will require some additional tinkering with XUL internals, I'll try to work on this in spare time

**Comment by lidel** [comment]
As for CORS, seems that a fix landed this week: https://github.com/ipfs/go-ipfs/pull/1529

**Comment by lidel** [comment]
Continued in https://github.com/lidel/ipfs-firefox-addon/issues/48
